java/ql/src/utils/stub-generator/Stubs.qll

```diff
@@ -404,20 +400,45 @@ private string stubMember(Member m) {
 
 private string stubAnnotation(Annotation a) {
   if exists(a.getAValue())
-  then result = a.toString() + "(" + concat(stubAnnotationValue(a.getAValue()), ",") + ")"
-  else result = a.toString()
+  then
+    result =
+      a.getType().getName() + "(" +
+        concat(string name, Expr value |
+          value = a.getValue(name)
+        |
+          name + "=" + stubAnnotationValue(value), ","
+        ) + ")"
+  else result = a.getType().getName()
 }
 
-private string stubAnnotiationSimpleValue(Expr value) {
-  result = value.(FieldAccess).getField().getQualifiedName() or
-  result = value.(Literal).toString()
+private string stubAnnotationSimpleValue(Expr value) {
+  result = value.(FieldAccess).getField().getQualifiedName()
+  or
+  result = value.(Literal).getLiteral()
+  or
+  not value instanceof Literal and
+  (
+    result = value.(CompileTimeConstantExpr).getStringValue() or
+    result = value.(CompileTimeConstantExpr).getBooleanValue().toString() or
+    result = value.(CompileTimeConstantExpr).getIntValue().toString()
+  )
+  or
+  result = value.(Annotation).getType().getName()
```

Comment:
Is this really correct? Shouldn't it be `result = stubAnnotation(value)`? `@Outer(@Inner)` was only a simple example, the annotations and their values can be nested arbitrarily deep, e.g. `@Outer(@Inner(@Inner2("some value")))`.

This would then require to add the `@` prefix in `subAnnotation` instead of in `stubAnnotations`.

Reply:
Yes, I thought of that, but using `stubAnnotation` there creates a non-monotonic recursion and CodeQL doesn't like it. I thought that we could at least support the basic case of a nested annotation, and if we end up needing to stub more complex cases we can think about this again.

Reply:
Thanks for the clarification, I had not tested the code myself. It probably still needs the `@` prefix then because that won't be included by `getName()`, right?

Do you think it would be useful to add a comment there describing the limitation you mentioned.

Reply:
Sure, did it in 373b64b.

Reply:
Using a `language[monotonicAggregates]` annotation on `stubAnnotation` would allow the recursion to be monotinic

Reply:
TIL, thanks 😄 Now it looks better: 1f1e494